// handpose/src/Proctor.js
// Import dependencies
import React, { useRef, useState, useEffect, useCallback } from "react";
import * as tf from "@tensorflow/tfjs";
import * as cocossd from "@tensorflow-models/coco-ssd";
import Webcam from "react-webcam";
import "./App.css";
import { drawRect } from "./utilities";
import { div } from "@tensorflow/tfjs";

function Proctor(props) {
  const [data, setData] = useState(null);
  const parentDiv = useRef();
  const webcamRef = useRef(null);
  const canvasRef = useRef(null);
  const photoRef = useRef(null);
  const app = useRef(null);

  useEffect(() => {}, [data]);

  useEffect(() => {
    runCoco();
  }, []);

  useEffect(() => {
    takePicture();
    setInterval(() => {
      takePicture();
    }, 5000);
  }, []);

  const takePicture = () => {
    let width = 300;
    let height = 300;
    let video = webcamRef.current.video;
    let photo = photoRef.current;
    photo.width = width;
    photo.height = height;
    const d = new Date();
    let ctx = photo.getContext("2d");
    ctx.drawImage(video, 0, 0, width, height);
    ctx.fillStyle = "black";
    ctx.font = "20px Arial";
    ctx.strokeText(
      `${d.getFullYear()}-${
        d.getMonth() + 1
      }-${d.getDate()}  ${d.getHours()}:${d.getMinutes()}:${d.getSeconds()}`,
      100,
      50
    );
    let photoToSend = photoRef.current.toDataURL("image/png");
    props.setImage(photoToSend);
  };

  const runCoco = async () => {
    const net = await cocossd.load();
    //  Loop and detect hands
    setInterval(() => {
      detect(net);
    }, 1000);
  };

  let countSeconds = 0;

  const detect = async (net) => {
    // Check data is available
    if (
      typeof webcamRef.current !== "undefined" &&
      webcamRef.current !== null &&
      webcamRef.current.video.readyState === 4
    ) {
      // Get Video Properties
      const video = webcamRef.current.video;
      const videoWidth = webcamRef.current.video.videoWidth;
      const videoHeight = webcamRef.current.video.videoHeight;

      // Set video width
      webcamRef.current.video.width = videoWidth;
      webcamRef.current.video.height = videoHeight;

      // Set canvas height and width
      canvasRef.current.width = videoWidth;
      canvasRef.current.height = videoHeight;

      // Make Detections
      const obj = await net.detect(video);
      setData(
        new Map([
          [obj.map((x) => x.class), obj.map((x) => x.score)],
          ["timeStamp", Date.now()],
        ])
      );

      let countPerson = 0;
      let personFound = false;
      for (var i = 0; i < obj.length; i++) {
        if (obj[i].class === "cell phone") {
          props.setMobilePhone(true);
          takePicture();
          break;
        }
        if (obj[i].class === "person") {
          personFound = true;
          countPerson++;
          if (countPerson >= 2) {
            takePicture();
            props.setIsMoreThanOnePerson(true);
            break;
          }
        }
      }
      if (!personFound) {
        countSeconds++;
        if (countSeconds >= 5) {
          props.setPersonMissing(true);
        }
      } else {
        countSeconds = 0;
      }

      // Draw mesh
      const ctx = canvasRef.current.getContext("2d");
      drawRect(obj, ctx);
    }
  };

  const printScreenFunction = useCallback((event) => {
    if (event.key === "PrintScreen") {
      console.log("Print Screen pressed");
    }
  }, []);

  useEffect(() => {
    document.addEventListener("keydown", printScreenFunction);
  }, [printScreenFunction]);

  return (
    <div ref={app}>
      <div>
        <div className="App" ref={parentDiv}>
          <Webcam
            screenshotFormat="image/jpeg"
            ref={webcamRef}
            muted={true}
            style={{
              position: "absolute",
              left: 0,
              bottom: 0,
              zindex: 9,
              width: 300,
              height: 300,
            }}
          />

          <canvas style={{ display: "none" }} ref={canvasRef} />
          <canvas style={{ display: "none" }} ref={photoRef}></canvas>
        </div>
      </div>
    </div>
  );
}

export default Proctor;
